### PR TITLE
kubernetes: Make cockpit-kube-auth work with 1.6 in RBAC mode

### PR DESCRIPTION
**Pull request by petervo**

### CODE REVIEW

**Comment by moon03432** [comment]
@petervo Do you have a docker image including that fix? I work with @hongtao12310, we are going to test it.

**Comment by petervo** [comment]
@moon03432, not publicly available no. On a fedora-25 machine you can checkout this repo (see: https://github.com/cockpit-project/cockpit/blob/master/HACKING.md) and run

```make kubernetes-container```

to build one yourself. To do this on a different machine change ```containers/kubernetes/Dockerfile``` to be based on a image that matches the machine you are building on.

File: test/containers/check-kubernetes [hunk 1]
```diff
@@ -51,25 +51,99 @@ class KubernetesContainerMachine(VirtMachine):
...
     machine_class = KubernetesContainerMachine
 
+    def get_bad_auth_code(self, port=6443):
+        code = self.machine.execute('curl -k -u bad:bad -s -o /dev/null -w "%{{http_code}}" https://{}:{}/api/namespaces/kubernetes'.format(self.machine.address, port))
+        return code.strip()
+
+    def restart_api_server(self, **kwargs):
+        # Delete the kubernetes service so that it will need to be recreated
+        # Do this because RBAC takes a while to finish setup
+        # and there is no good way to track it's progress
```
Comment (martinpitt) [inline test/containers/check-kubernetes]: typo (sorry, OCD): "its"

**Comment by hongtao12310** [comment]
@petervo I built a new cockpit docker images based on your change. sounds like there is something wrong after apply this patch.


[root@ctum2a2402005 cockpit]# kubectl logs kubernetes-cockpit-vr16v
DEBUG: cockpit-ws: loaded 1 certificates from /etc/cockpit/ws-certs.d/0-self-signed.cert
INFO: cockpit-ws: Using certificate: /etc/cockpit/ws-certs.d/0-self-signed.cert
DEBUG: cockpit-protocol: Loaded configuration from: /etc/cockpit/cockpit.conf
DEBUG: cockpit-ws: auth is idle


I can't even execute command inside the cockpit container.
[root@ctum2a2402005 cockpit]# kubectl exec kubernetes-cockpit-vr16v ls /
rpc error: code = 13 desc = invalid header field value "oci runtime error: exec failed: cannot exec a container that has run and stopped\n"

**Comment by petervo** [comment]
@hongtao12310 there's not enough information here for me to tell what is wrong. But changes are it's a version mismatch between cockpit-ws installed in the container and the built auth libraries. Can you share your container with me?

File: pkg/kubernetes/standalone/src/cockpit-kube-auth/helpers/client.go [hunk 2]
```diff
@@ -133,35 +133,63 @@ func (self *Client) fetchVersion(authHeader string) error {
...
+		status, e = self.apiStatus("namespaces", "Basic")
+
+		// Some versions of kubernetes return 403 instead of 401
+		// when presented with bad basic auth data. In those cases
+		// we need to refuse authentication.
+		// https://github.com/kubernetes/kubernetes/pull/41775
+		if status == 403 {
+			e = errors.New("This version of kubernetes is not supported. Turn off anonymous auth or upgrade.")
+		} else if status == 200 || status == 401 {
+			success = true
```
Comment (petervo) [inline pkg/kubernetes/standalone/src/cockpit-kube-auth/helpers/client.go]: There are multiple calls happening here, it's not all the same. The comments describe the steps i'll try to make them clearer. But it's not simple to follow.

**Comment by petervo** [comment]
@martinpitt updated the comments, make more sense to you now?

**Comment by hongtao12310** [comment]
@petervo I just clone your repo and check out the kube-auth branch . and RUN make kubernetes-container to build the image.

and the build server is centos 7.2 based.

let me know if you need more details.

**Comment by petervo** [comment]
@hongtao12310 did you change the docker file to use a centos base image? Otherwise that won't work.

File: pkg/kubernetes/standalone/src/cockpit-kube-auth/helpers/client.go [hunk 3]
```diff
@@ -133,35 +133,63 @@ func (self *Client) fetchVersion(authHeader string) error {
...
+	// We need to determine if that is because no auth is configured
+	// Or if basic auth is broken.
+	if status == 200 || status == 403 {
+		// Send an empty basic header for namespaces that is guarenteed to be invalid.
+		// This should return a 200 if the api is open or a 401 if the
+		// api is protected.
+		status, e = self.apiStatus("namespaces", "Basic")
+
+		// Some versions of kubernetes return 403 instead of 401
+		// when presented with bad basic auth data. In those cases
```
Comment (petervo) [inline pkg/kubernetes/standalone/src/cockpit-kube-auth/helpers/client.go]: here we sent it bad auth data on purpose

**Comment by petervo** [comment]
@hongtao12310 I'm pushing a build here for you to try.

https://hub.docker.com/r/petervo/test/

LMK how it goes.

File: pkg/kubernetes/standalone/src/cockpit-kube-auth/helpers/client.go (hunk 2 above)
Comment (martinpitt) [inline pkg/kubernetes/standalone/src/cockpit-kube-auth/helpers/client.go]: Oh right, this `status` is for the invalid "namespaces" call while the previous `status` was for the "" call. Thanks for adding the comment, that helps to understand it!

**Comment by martinpitt** [comment]
Should we wait for @hongtao12310 's testing feedback for landing this? If not, please feel free to land yourself.

**Comment by hongtao12310** [comment]
@petervo 

just tried your docker images. 

cockpit said that "Authentication failed: internal-error: This version of kubernetes is not supported. Turn off anonymous auth or upgrade"

should I set --anonymous-auth=false when boot up the APIServer ?

**Comment by petervo** [comment]
@hongtao12310, yes. Basically with anonymous auth turned on and without this fix https://github.com/kubernetes/kubernetes/pull/41775. There is no way to tell the difference between a good u/p but access denied by configuration and a bad u/p. So we can't log users like that in.

So you have 2 options, run kube api server with ```--anonymous-auth=false``` or if you really don't care about auth at all is to run

```
kubectl create clusterrolebinding anon-cluster-admin-binding --clusterrole=cluster-admin --user=system:anonymous
```

That grants everything to system:anonymous and now cockpit won't even prompt for a u/p. I do not recommend doing this for anything more than a toy/POC setup.

**Pull request closed without merging**